addons/account/models/chart_template.py

```diff
@@ -30,6 +30,97 @@ def preserve_existing_tags_on_taxes(cr, registry, module):
 
+def update_taxes_from_templates(cr, chart_template_xmlid):
+    def _create_tax_from_template(company, template, old_tax=None):
+        """
+        Create a new tax from template with template xmlid, if there was already an old tax with that xmlid we
+        remove the xmlid from it but don't modify anything else.
+        """
+        def _remove_xml_id(xml_id):
+            module, name = xml_id.split(".", 1)
+            env['ir.model.data'].search([('module', '=', module), ('name', '=', name)]).unlink()
+
+        template_vals = template._get_tax_vals_complete(company)
+        chart_template = env["account.chart.template"].with_context(default_company_id=company.id)
+        if old_tax:
+            xml_id = old_tax.get_xml_id().get(old_tax.id)
+            if xml_id:
+                _remove_xml_id(xml_id)
+        chart_template.create_record_with_xmlid(company, template, "account.tax", template_vals)
+
+    def _get_template_to_tax_xmlid_mapping(company):
+        """
+        This function uses ir_model_data to return a mapping between the tax templates and the taxes, using their xmlid
+        :returns: {
+            account.tax.template.id: account.tax.id
+            }
+        """
+        env['ir.model.data'].flush()
+        env.cr.execute(
+            """
+            SELECT template.res_id AS template_res_id,
+                   tax.res_id AS tax_res_id
+            FROM ir_model_data tax
+            JOIN ir_model_data template
+            ON template.name = substr(tax.name, strpos(tax.name, '_') + 1)
+            WHERE tax.model = 'account.tax'
+            AND tax.name LIKE %s
+            -- tax.name is of the form: {company_id}_{account.tax.template.name}
+            """,
+            [r"%s\_%%" % company.id],
+        )
+        tuples = env.cr.fetchall()
+        return dict(tuples)
+
+    def _is_tax_and_template_same(template, tax):
+        """
+        This function compares account.tax and account.tax.template repartition lines.
+        A tax is considered the same as the template if they have the same:
+            - amount_type
+            - amount
+            - repartition lines percentages in the same order
+        """
+        if tax.amount_type != template.amount_type or tax.amount != template.amount \
+                or len(tax.invoice_repartition_line_ids) != len(template.invoice_repartition_line_ids) \
+                or len(tax.refund_repartition_line_ids) != len(template.refund_repartition_line_ids):
+            return False
+
+        tax_rep_lines = tax.invoice_repartition_line_ids + tax.refund_repartition_line_ids
+        template_rep_lines = template.invoice_repartition_line_ids + template.refund_repartition_line_ids
+        for rep_line_tax, rep_line_template in zip(tax_rep_lines, template_rep_lines):
+            if rep_line_tax.factor_percent != rep_line_template.factor_percent:
+                return False
+        return True
```

Comment:
One single return statement is easier to read than multiple with various logic.
Also, PEP8 prefers parenthesis to `\` to break long lines

```suggestion
        tax_rep_lines = tax.invoice_repartition_line_ids + tax.refund_repartition_line_ids
        template_rep_lines = template.invoice_repartition_line_ids + template.refund_repartition_line_ids
        return (
            tax.amount_type == template.amount_type
            and tax.amount == template.amount
            and len(tax_rep_lines) == len(template_rep_lines)
            and all(
                rep_line_tax.factor_percent == rep_line_template.factor_percent
                for rep_line_tax, rep_line_template in zip(tax_rep_lines, template_rep_lines)
            )
        )
```